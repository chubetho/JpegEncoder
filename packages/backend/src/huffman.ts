import type { Buffer } from 'node:buffer'
import { sort } from 'fast-sort'

interface Item { symbol?: number, count: number, isPackage?: boolean }
interface LengthBookItem { symbol: number, length: number }
export interface Node {
  symbol?: number
  left?: Node
  right?: Node
}

export function computeLengthBook(buffer: Buffer, maxLength = 15) {
  const hg = new Map<number, number>()

  buffer.forEach(x => hg.set(x, (hg.get(x) ?? 0) + 1))

  if (hg.size < 2 || Math.ceil(Math.log2(hg.size)) > maxLength)
    return []

  const originalRow: Item[] = sort(
    Array.from(hg).map(([k, v]) => ({ symbol: k, count: v })),
  ).asc(i => i.count)
  const originalLength = originalRow.length

  const table: Item[][] = [originalRow]
  for (let i = 0; i < maxLength - 1; i++)
    table.push(computeRow(table[i], originalRow))

  const codeLength = Array.from(({ length: originalLength })).fill(0) as number[]
  const numProcess = [originalLength * 2 - 2]
  let symbolCount = 0
  let mergedCount = 0

  for (let i = table.length - 1; i >= 0; i--) {
    const row = [...table[i]]
    row.splice(numProcess[table.length - 1 - i])
    row.forEach((x) => {
      if (x.isPackage) {
        mergedCount += 1
      }
      else {
        codeLength[symbolCount] += 1
        symbolCount += 1
      }
    })
    numProcess.push(2 * mergedCount)
    symbolCount = 0
    mergedCount = 0
  }

  const merged = originalRow.reduce((acc, curr, index) => {
    curr.symbol && acc.push({ symbol: curr.symbol, length: codeLength[index] })
    return acc
  }, [] as { symbol: number, length: number }[])

  const check = merged.map(x => x.length).reduce((acc, curr) => acc + 2 ** -curr, 0) === 1

  return check ? merged : []
}

function computeRow(arr: Item[], original: Item[]) {
  const row = [...original]
  arr.forEach((item, index) => {
    if ((index & 1) !== 0)
      return

    const left = arr.at(index)
    if (!left)
      return

    const right = arr.at(index + 1)
    if (!right)
      return

    row.push({
      count: left.count + right.count,
      isPackage: true,
    })
  })
  return sort(row).asc(i => i.count)
}

export function computeCodeBook(lengthBook: LengthBookItem[], maxLength = 15, anti = true) {
  if (!lengthBook.length)
    return {}

  let code = 0
  const codeBook: Record<number, string> = {}

  const shortestIndex = lengthBook.length - 1
  const shortestIndexLength = lengthBook[shortestIndex].length
  codeBook[lengthBook[shortestIndex].symbol] = ''.padStart(shortestIndexLength, '0')

  for (let i = lengthBook.length - 2; i >= 0; i--) {
    code += 1
    if (lengthBook[i].length === shortestIndexLength) {
      codeBook[lengthBook[i].symbol] = code.toString(2).padStart(shortestIndexLength, '0')
      continue
    }

    while (code.toString(2).length < lengthBook[i].length)
      code <<= 1

    codeBook[lengthBook[i].symbol] = code.toString(2)
  }

  if (anti) {
    const target = Object.entries(codeBook)
      .find(([_, value]) => value.split('').every(char => char === '1'))

    if (target && target[0].length + 1 <= maxLength)
      codeBook[+target[0]] = `${target[1]}0`
  }

  return codeBook
}

export function computeCharTable() {
  const table: string[] = Array.from({ length: 256 })
  const fn = String.fromCharCode
  for (let i = 0; i < 256; i++)
    table[i] = fn(i)

  return table
}
