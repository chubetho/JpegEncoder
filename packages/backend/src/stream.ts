import * as fs from 'node:fs'

export function useStream() {
  let buffer = new Uint8Array(1024)
  let byte = 0
  let bitCount = 0
  let bytePos = 0

  function writeBit(bit: number) {
    if (bit !== 0 && bit !== 1)
      throw new Error('Invalid value for bit')

    byte = (byte << 1) | bit
    bitCount++

    if (bitCount === 8) {
      buffer[bytePos++] = byte
      reset()

      checkBuffer()
    }
  }

  function readBit() {
    if (bytePos === 0 && bitCount === 0 && byte === 0)
      throw new Error('Buffer is empty')

    if (bitCount === 0) {
      bytePos--
      byte = buffer[bytePos]
      bitCount = 8
    }

    bitCount--
    return (byte >> bitCount) & 1
  }

  function save(path: string) {
    if (bitCount && byte) {
      buffer[bytePos] = byte
      bytePos++
      reset()
    }

    fs.writeFileSync(path, buffer.subarray(0, bytePos))
  }

  function writeByte(v: number) {
    if (!Number.isInteger(v) || v < 0 || v > 255)
      throw new Error('Value must be a 8 bit integer')

    buffer[bytePos++] = v
    checkBuffer()
  }

  function writeWord(v: number) {
    if (!Number.isInteger(v) || v < 0 || v > 65535)
      throw new Error('Value must be a 16 bit integer')

    writeByte((v >> 8) & 0xFF)
    writeByte(v & 0xFF)
  }

  function checkBuffer() {
    if (bytePos < buffer.length)
      return

    const newBuffer = new Uint8Array(buffer.length * 2)
    newBuffer.set(buffer)
    buffer = newBuffer
  }

  function reset() {
    byte = 0
    bitCount = 0
  }

  function getBuffer() {
    return buffer
  }

  return { writeBit, writeByte, writeWord, readBit, save, getBuffer }
}
