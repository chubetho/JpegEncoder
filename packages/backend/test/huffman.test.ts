import { describe, expect, it } from 'vitest'
import { computeBinaryMap, computeTree, useHuffman } from '../src/huffman.js'
import { useStream } from '../src/stream.js'

const str = 'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Fugit reiciendis perspiciatis expedita nemo a libero, quas in maiores placeat. Dolore quam ipsa iure fugiat veniam? Blanditiis magni eveniet minus nesciunt ea corrupti optio rerum obcaecati distinctio maxime deleniti, quo molestiae unde est, tempore ad labore dolorum cum tempora consectetur quia incidunt facilis. Fugiat est magnam fuga cumque, architecto pariatur asperiores ad qui similique, cum tenetur laboriosam sed placeat ullam, omnis nulla. Tenetur rerum molestiae quod esse labore consequatur laudantium officiis aliquam. Consequatur veritatis saepe nisi, unde voluptatibus at sequi adipisci. Maxime officiis exercitationem id aliquam ullam iusto fugiat autem laudantium?'

describe('huffmann', () => {
  it('compute tree', () => {
    const root = computeTree(str)
    expect(root).toMatchInlineSnapshot(`
      {
        "count": 739,
        "left": {
          "count": 300,
          "left": {
            "count": 139,
            "left": {
              "count": 67,
              "symbol": "e",
            },
            "right": {
              "count": 72,
              "left": {
                "count": 35,
                "symbol": "m",
              },
              "right": {
                "count": 37,
                "symbol": "o",
              },
            },
          },
          "right": {
            "count": 161,
            "left": {
              "count": 80,
              "symbol": "i",
            },
            "right": {
              "count": 81,
              "left": {
                "count": 39,
                "symbol": "s",
              },
              "right": {
                "count": 42,
                "left": {
                  "count": 19,
                  "symbol": "d",
                },
                "right": {
                  "count": 23,
                  "left": {
                    "count": 10,
                    "left": {
                      "count": 4,
                      "left": {
                        "count": 2,
                        "left": {
                          "count": 1,
                          "symbol": "B",
                        },
                        "right": {
                          "count": 1,
                          "symbol": "h",
                        },
                      },
                      "right": {
                        "count": 2,
                        "left": {
                          "count": 1,
                          "symbol": "T",
                        },
                        "right": {
                          "count": 1,
                          "symbol": "C",
                        },
                      },
                    },
                    "right": {
                      "count": 6,
                      "symbol": ".",
                    },
                  },
                  "right": {
                    "count": 13,
                    "symbol": "q",
                  },
                },
              },
            },
          },
        },
        "right": {
          "count": 439,
          "left": {
            "count": 199,
            "left": {
              "count": 99,
              "symbol": " ",
            },
            "right": {
              "count": 100,
              "left": {
                "count": 50,
                "symbol": "u",
              },
              "right": {
                "count": 50,
                "symbol": "t",
              },
            },
          },
          "right": {
            "count": 240,
            "left": {
              "count": 112,
              "left": {
                "count": 53,
                "left": {
                  "count": 26,
                  "symbol": "c",
                },
                "right": {
                  "count": 27,
                  "symbol": "l",
                },
              },
              "right": {
                "count": 59,
                "left": {
                  "count": 29,
                  "left": {
                    "count": 13,
                    "left": {
                      "count": 6,
                      "symbol": "b",
                    },
                    "right": {
                      "count": 7,
                      "left": {
                        "count": 3,
                        "left": {
                          "count": 1,
                          "symbol": "M",
                        },
                        "right": {
                          "count": 2,
                          "symbol": "F",
                        },
                      },
                      "right": {
                        "count": 4,
                        "symbol": "x",
                      },
                    },
                  },
                  "right": {
                    "count": 16,
                    "left": {
                      "count": 8,
                      "symbol": ",",
                    },
                    "right": {
                      "count": 8,
                      "symbol": "g",
                    },
                  },
                },
                "right": {
                  "count": 30,
                  "symbol": "n",
                },
              },
            },
            "right": {
              "count": 128,
              "left": {
                "count": 62,
                "symbol": "a",
              },
              "right": {
                "count": 66,
                "left": {
                  "count": 33,
                  "symbol": "r",
                },
                "right": {
                  "count": 33,
                  "left": {
                    "count": 16,
                    "left": {
                      "count": 8,
                      "symbol": "f",
                    },
                    "right": {
                      "count": 8,
                      "left": {
                        "count": 4,
                        "symbol": "v",
                      },
                      "right": {
                        "count": 4,
                        "left": {
                          "count": 2,
                          "symbol": "?",
                        },
                        "right": {
                          "count": 2,
                          "left": {
                            "count": 1,
                            "symbol": "L",
                          },
                          "right": {
                            "count": 1,
                            "symbol": "D",
                          },
                        },
                      },
                    },
                  },
                  "right": {
                    "count": 17,
                    "symbol": "p",
                  },
                },
              },
            },
          },
        },
      }
    `)
  })

  it('compute binary map', () => {
    const root = computeTree(str)
    const { map, level } = computeBinaryMap(root)
    expect(level).toEqual(10)
    expect(map).toMatchSnapshot()
  })

  it('useHuffman with invalid input', () => {
    expect(() => useHuffman('')).toThrowErrorMatchingInlineSnapshot('"Invalid input"')
  })

  it('encode', () => {
    const { encode, encoded } = useHuffman(str)
    encode()
    expect(encoded).toMatchSnapshot()
  })

  it('decode', () => {
    const { encode, decode } = useHuffman(str)
    expect(() => decode()).toThrowErrorMatchingInlineSnapshot('"encode function must be called first"')

    encode()
    expect(decode()).toEqual(str)
  })

  it('buffer', () => {
    const { encode, encoded, level } = useHuffman(str)
    encode()

    const { getBuffer, writeWord } = useStream()

    encoded.forEach((symbol) => {
      const byte = symbol.padStart(16, '1')
      writeWord(Number.parseInt(byte, 2))
    })
    console.log(getBuffer())
  })
})
