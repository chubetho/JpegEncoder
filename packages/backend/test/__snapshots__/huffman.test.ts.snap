// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`huffmann > compute binary map 1`] = `
Map {
  "e" => "000",
  "m" => "0010",
  "o" => "0011",
  "i" => "010",
  "s" => "0110",
  "d" => "01110",
  "B" => "011110000",
  "h" => "011110001",
  "T" => "011110010",
  "C" => "011110011",
  "." => "0111101",
  "q" => "011111",
  " " => "100",
  "u" => "1010",
  "t" => "1011",
  "c" => "11000",
  "l" => "11001",
  "b" => "1101000",
  "M" => "110100100",
  "F" => "110100101",
  "x" => "11010011",
  "," => "1101010",
  "g" => "1101011",
  "n" => "11011",
  "a" => "1110",
  "r" => "11110",
  "f" => "1111100",
  "v" => "11111010",
  "?" => "111110110",
  "L" => "1111101110",
  "D" => "1111101111",
  "p" => "111111",
}
`;

exports[`huffmann > encode 1`] = `
[
  "1111101110",
  "0011",
  "11110",
  "000",
  "0010",
  "100",
  "010",
  "111111",
  "0110",
  "1010",
  "0010",
  "100",
  "01110",
  "0011",
  "11001",
  "0011",
  "11110",
  "100",
  "0110",
  "010",
  "1011",
  "100",
  "1110",
  "0010",
  "000",
  "1011",
  "1101010",
  "100",
  "11000",
  "0011",
  "11011",
  "0110",
  "000",
  "11000",
  "1011",
  "000",
  "1011",
  "1010",
  "11110",
  "100",
  "1110",
  "01110",
  "010",
  "111111",
  "010",
  "0110",
  "010",
  "11000",
  "010",
  "11011",
  "1101011",
  "100",
  "000",
  "11001",
  "010",
  "1011",
  "0111101",
  "100",
  "110100101",
  "1010",
  "1101011",
  "010",
  "1011",
  "100",
  "11110",
  "000",
  "010",
  "11000",
  "010",
  "000",
  "11011",
  "01110",
  "010",
  "0110",
  "100",
  "111111",
  "000",
  "11110",
  "0110",
  "111111",
  "010",
  "11000",
  "010",
  "1110",
  "1011",
  "010",
  "0110",
  "100",
  "000",
  "11010011",
  "111111",
  "000",
  "01110",
  "010",
  "1011",
  "1110",
  "100",
  "11011",
  "000",
  "0010",
  "0011",
  "100",
  "1110",
  "100",
  "11001",
  "010",
  "1101000",
  "000",
  "11110",
  "0011",
  "1101010",
  "100",
  "011111",
  "1010",
  "1110",
  "0110",
  "100",
  "010",
  "11011",
  "100",
  "0010",
  "1110",
  "010",
  "0011",
  "11110",
  "000",
  "0110",
  "100",
  "111111",
  "11001",
  "1110",
  "11000",
  "000",
  "1110",
  "1011",
  "0111101",
  "100",
  "1111101111",
  "0011",
  "11001",
  "0011",
  "11110",
  "000",
  "100",
  "011111",
  "1010",
  "1110",
  "0010",
  "100",
  "010",
  "111111",
  "0110",
  "1110",
  "100",
  "010",
  "1010",
  "11110",
  "000",
  "100",
  "1111100",
  "1010",
  "1101011",
  "010",
  "1110",
  "1011",
  "100",
  "11111010",
  "000",
  "11011",
  "010",
  "1110",
  "0010",
  "111110110",
  "100",
  "011110000",
  "11001",
  "1110",
  "11011",
  "01110",
  "010",
  "1011",
  "010",
  "010",
  "0110",
  "100",
  "0010",
  "1110",
  "1101011",
  "11011",
  "010",
  "100",
  "000",
  "11111010",
  "000",
  "11011",
  "010",
  "000",
  "1011",
  "100",
  "0010",
  "010",
  "11011",
  "1010",
  "0110",
  "100",
  "11011",
  "000",
  "0110",
  "11000",
  "010",
  "1010",
  "11011",
  "1011",
  "100",
  "000",
  "1110",
  "100",
  "11000",
  "0011",
  "11110",
  "11110",
  "1010",
  "111111",
  "1011",
  "010",
  "100",
  "0011",
  "111111",
  "1011",
  "010",
  "0011",
  "100",
  "11110",
  "000",
  "11110",
  "1010",
  "0010",
  "100",
  "0011",
  "1101000",
  "11000",
  "1110",
  "000",
  "11000",
  "1110",
  "1011",
  "010",
  "100",
  "01110",
  "010",
  "0110",
  "1011",
  "010",
  "11011",
  "11000",
  "1011",
  "010",
  "0011",
  "100",
  "0010",
  "1110",
  "11010011",
  "010",
  "0010",
  "000",
  "100",
  "01110",
  "000",
  "11001",
  "000",
  "11011",
  "010",
  "1011",
  "010",
  "1101010",
  "100",
  "011111",
  "1010",
  "0011",
  "100",
  "0010",
  "0011",
  "11001",
  "000",
  "0110",
  "1011",
  "010",
  "1110",
  "000",
  "100",
  "1010",
  "11011",
  "01110",
  "000",
  "100",
  "000",
  "0110",
  "1011",
  "1101010",
  "100",
  "1011",
  "000",
  "0010",
  "111111",
  "0011",
  "11110",
  "000",
  "100",
  "1110",
  "01110",
  "100",
  "11001",
  "1110",
  "1101000",
  "0011",
  "11110",
  "000",
  "100",
  "01110",
  "0011",
  "11001",
  "0011",
  "11110",
  "1010",
  "0010",
  "100",
  "11000",
  "1010",
  "0010",
  "100",
  "1011",
  "000",
  "0010",
  "111111",
  "0011",
  "11110",
  "1110",
  "100",
  "11000",
  "0011",
  "11011",
  "0110",
  "000",
  "11000",
  "1011",
  "000",
  "1011",
  "1010",
  "11110",
  "100",
  "011111",
  "1010",
  "010",
  "1110",
  "100",
  "010",
  "11011",
  "11000",
  "010",
  "01110",
  "1010",
  "11011",
  "1011",
  "100",
  "1111100",
  "1110",
  "11000",
  "010",
  "11001",
  "010",
  "0110",
  "0111101",
  "100",
  "110100101",
  "1010",
  "1101011",
  "010",
  "1110",
  "1011",
  "100",
  "000",
  "0110",
  "1011",
  "100",
  "0010",
  "1110",
  "1101011",
  "11011",
  "1110",
  "0010",
  "100",
  "1111100",
  "1010",
  "1101011",
  "1110",
  "100",
  "11000",
  "1010",
  "0010",
  "011111",
  "1010",
  "000",
  "1101010",
  "100",
  "1110",
  "11110",
  "11000",
  "011110001",
  "010",
  "1011",
  "000",
  "11000",
  "1011",
  "0011",
  "100",
  "111111",
  "1110",
  "11110",
  "010",
  "1110",
  "1011",
  "1010",
  "11110",
  "100",
  "1110",
  "0110",
  "111111",
  "000",
  "11110",
  "010",
  "0011",
  "11110",
  "000",
  "0110",
  "100",
  "1110",
  "01110",
  "100",
  "011111",
  "1010",
  "010",
  "100",
  "0110",
  "010",
  "0010",
  "010",
  "11001",
  "010",
  "011111",
  "1010",
  "000",
  "1101010",
  "100",
  "11000",
  "1010",
  "0010",
  "100",
  "1011",
  "000",
  "11011",
  "000",
  "1011",
  "1010",
  "11110",
  "100",
  "11001",
  "1110",
  "1101000",
  "0011",
  "11110",
  "010",
  "0011",
  "0110",
  "1110",
  "0010",
  "100",
  "0110",
  "000",
  "01110",
  "100",
  "111111",
  "11001",
  "1110",
  "11000",
  "000",
  "1110",
  "1011",
  "100",
  "1010",
  "11001",
  "11001",
  "1110",
  "0010",
  "1101010",
  "100",
  "0011",
  "0010",
  "11011",
  "010",
  "0110",
  "100",
  "11011",
  "1010",
  "11001",
  "11001",
  "1110",
  "0111101",
  "100",
  "011110010",
  "000",
  "11011",
  "000",
  "1011",
  "1010",
  "11110",
  "100",
  "11110",
  "000",
  "11110",
  "1010",
  "0010",
  "100",
  "0010",
  "0011",
  "11001",
  "000",
  "0110",
  "1011",
  "010",
  "1110",
  "000",
  "100",
  "011111",
  "1010",
  "0011",
  "01110",
  "100",
  "000",
  "0110",
  "0110",
  "000",
  "100",
  "11001",
  "1110",
  "1101000",
  "0011",
  "11110",
  "000",
  "100",
  "11000",
  "0011",
  "11011",
  "0110",
  "000",
  "011111",
  "1010",
  "1110",
  "1011",
  "1010",
  "11110",
  "100",
  "11001",
  "1110",
  "1010",
  "01110",
  "1110",
  "11011",
  "1011",
  "010",
  "1010",
  "0010",
  "100",
  "0011",
  "1111100",
  "1111100",
  "010",
  "11000",
  "010",
  "010",
  "0110",
  "100",
  "1110",
  "11001",
  "010",
  "011111",
  "1010",
  "1110",
  "0010",
  "0111101",
  "100",
  "011110011",
  "0011",
  "11011",
  "0110",
  "000",
  "011111",
  "1010",
  "1110",
  "1011",
  "1010",
  "11110",
  "100",
  "11111010",
  "000",
  "11110",
  "010",
  "1011",
  "1110",
  "1011",
  "010",
  "0110",
  "100",
  "0110",
  "1110",
  "000",
  "111111",
  "000",
  "100",
  "11011",
  "010",
  "0110",
  "010",
  "1101010",
  "100",
  "1010",
  "11011",
  "01110",
  "000",
  "100",
  "11111010",
  "0011",
  "11001",
  "1010",
  "111111",
  "1011",
  "1110",
  "1011",
  "010",
  "1101000",
  "1010",
  "0110",
  "100",
  "1110",
  "1011",
  "100",
  "0110",
  "000",
  "011111",
  "1010",
  "010",
  "100",
  "1110",
  "01110",
  "010",
  "111111",
  "010",
  "0110",
  "11000",
  "010",
  "0111101",
  "100",
  "110100100",
  "1110",
  "11010011",
  "010",
  "0010",
  "000",
  "100",
  "0011",
  "1111100",
  "1111100",
  "010",
  "11000",
  "010",
  "010",
  "0110",
  "100",
  "000",
  "11010011",
  "000",
  "11110",
  "11000",
  "010",
  "1011",
  "1110",
  "1011",
  "010",
  "0011",
  "11011",
  "000",
  "0010",
  "100",
  "010",
  "01110",
  "100",
  "1110",
  "11001",
  "010",
  "011111",
  "1010",
  "1110",
  "0010",
  "100",
  "1010",
  "11001",
  "11001",
  "1110",
  "0010",
  "100",
  "010",
  "1010",
  "0110",
  "1011",
  "0011",
  "100",
  "1111100",
  "1010",
  "1101011",
  "010",
  "1110",
  "1011",
  "100",
  "1110",
  "1010",
  "1011",
  "000",
  "0010",
  "100",
  "11001",
  "1110",
  "1010",
  "01110",
  "1110",
  "11011",
  "1011",
  "010",
  "1010",
  "0010",
  "111110110",
]
`;
